### PR TITLE
[PORT] View audit

### PR DESCRIPTION
**Pull request by jlsnow301**
## About The Pull Request
Ports an audit from BeeStation by @park66665
See [#3431](https://github.com/BeeStation/BeeStation-Hornet/pull/3431)
- Nothing user facing
- Might need kinks worked out
- Has been live for nearly a year downstream

## Why It's Good For The Game

Contains a huge number of performance fixes for the game
## Changelog

N/A nothing user facing

### CODE REVIEW

**Review by Ghommie — COMMENTED** [review]
This is not an exhaustive review, but the PR has quite a few compile errors that ought to have been fixed before commiting.

**Comment by jlsnow301** [comment]
> This is not an exhaustive review, but the PR has quite a few compile errors that ought to have been fixed before commiting.

good catch, thank you for looking through this. long road ahead for this PR

File: code/__DEFINES/icon_smoothing.dm [hunk 1]
```diff
@@ -26,7 +26,7 @@ DEFINE_BITFIELD(smoothing_flags, list(
 
 #define QUEUE_SMOOTH(thing_to_queue) if(thing_to_queue.smoothing_flags & (SMOOTH_CORNERS|SMOOTH_BITMASK)) {SSicon_smooth.add_to_queue(thing_to_queue)}
 
-#define QUEUE_SMOOTH_NEIGHBORS(thing_to_queue) for(var/neighbor in orange(1, thing_to_queue)) {var/atom/atom_neighbor = neighbor; QUEUE_SMOOTH(atom_neighbor)}
+#define QUEUE_SMOOTH_NEIGHBORS(thing_to_queue) for(var/atom/neighbor as anything in orange(1, thing_to_queue)) {if(neighbor.smooth) QUEUE_SMOOTH(neighbor)}
```
Comment (LemonInTheDark) [inline code/__DEFINES/icon_smoothing.dm:29]: What the heck is going on here

File: code/game/objects/structures/spirit_board.dm [hunk 2]
```diff
@@ -65,7 +65,7 @@
 
 	//mobs in range check
 	var/users_in_range = 0
-	for(var/mob/living/L in orange(1,src))
+	for(var/mob/living/L in oviewers(1,src))
```
Comment (LemonInTheDark) [inline code/game/objects/structures/spirit_board.dm:68]: This is a micro op, and not one that's sustainable to enforce. Why do it?

**Comment by LemonInTheDark** [comment]
Bee did this because they thought they didn't have any other source of cost. A lot of these changes are micro optimizations, some are just renames and some straight change behavior. The view() family causes problems not because people are using range(src, 1) instead of view(src, 1) or something silly, but because they will either do stupid chaining of large view procs in a way that ends up wasting time, or just really attempt to poll too large a range too often. If you act otherwise you'll waste your time finding them all, and the time of others who attempt to mimic you when feature coders inevitably fuck everything up.

Getting rid of range(0) stuff is fine mind, that's just old. But micro oping between range_turfs(1) - get_turf(src) and orange(1) is just kinda silly at some point, places undue work on the codebase and you. Unless it's in code hot enough to matter.

You should focus on hitting large ranges over the head I think, unless you can provide profilable differences significant for such low sizes and call counts. This sort of thing comes to mind https://github.com/tgstation/tgstation/pull/59980

Additionally if you're going to go through this work I'd prefer if you ported xeno's state of mind rather then their code, because bee's codebase and ours work differently. I've already found some behavior changes, and I don't like that.

File: code/__DEFINES/icon_smoothing.dm (hunk 1 above)
Comment (jlsnow301) [reply code/__DEFINES/icon_smoothing.dm:29]: not sure if i can use an if statement here but it is basically ```for(var/atom/neighbor as() in orange(1,A))
		if(neighbor.smooth)
			queue_smooth(neighbor)```

Comment (LemonInTheDark) [reply code/__DEFINES/icon_smoothing.dm:29]: why? that's not consistent with our current behavior

Comment (LemonInTheDark) [reply code/__DEFINES/icon_smoothing.dm:29]: you can by the way, though if you want newlines you need to use {} and ;

Comment (jlsnow301) [reply code/__DEFINES/icon_smoothing.dm]: reverted

**Comment by jlsnow301** [comment]
> Additionally if you're going to go through this work I'd prefer if you ported xeno's state of mind rather then their code, because bee's codebase and ours work differently. I've already found some behavior changes, and I don't like that.

going to scan over it a bit to find any game behavior changes - this should remain performance (even micro op) changes only

**Comment by LemonInTheDark** [comment]
I don't want to set a standard for micro ops like this, unless it's not actually a micro op. We have larger performance issues that it would be better to spend time on. Why is it worth doing if the effect is so small it's not noticeable?

**Comment by jlsnow301** [comment]
> I don't want to set a standard for micro ops like this, unless it's not actually a micro op. We have larger performance issues that it would be better to spend time on. Why is it worth doing if the effect is so small it's not noticeable?

i'm not sure how this would set a standard. is there any part i could remove that's bothering you? or is it just micro ops in general

**Comment by park66665** [comment]
Considering a psychiatric hospitalization happened in relation to the
original PR I certainly don't recommend doing the same with tg except
dealing with a few hot and ridiculous ones

**Comment by jlsnow301** [comment]
> Considering a psychiatric hospitalization happened in relation to the original PR I certainly don't recommend doing the same with tg except dealing with a few hot and ridiculous ones

Sure thing, closing this as per recommendations 
Just noting that I typed this all and it's passing checks ;_;

**Comment by park66665** [comment]
Holy moly I understand that as the original PR author my opinion actually
does matter but I wasn't telling you to close this; without doubt it is up
to you (you are also free to close this obviously)

**Comment by LemonInTheDark** [comment]
> i'm not sure how this would set a standard. is there any part i could remove that's bothering you? or is it just micro ops in general

It's like this.
You. me, the other coders and maints only have so much time and mental energy. There are things that are useful to direct that energy towards, things that make the code quality better, more resistant or useful, things that make the game more "FUN" (ss13 is a psy op don't let anyone tell you otherwise), things that make it look/feel/sound nicer (Optimization lies in here). Right?

We all only have so much time to work or maintain those things. There can be times where hell projects happen, huge time sinks. My harddel prs, action and style tguifying the codebase over a summer, wallening, etc. Given healthy patterns they can come with big benefits, but It's not great to do those things at high intensity or too much, because it burns you out and makes your life worse. Trust me trudging through a 1000 line conversion or 100 minor fixes with no break period is literal actual hell.

Performance is in that list of things. Note, normally actual expensive things come less from micro optmizations and more from changing how a system does the "work" involved. See kyler's hearing pr #61422 

This is mirrored in other aspects of the codebase too mind. #55595 and #50608 both come to mind. Raw time often does not directly correlate with return.

The microscopic benefits of range(1) returning slightly more objects in a corner then view(1) pales in comparison to the time you and I would have to spend changing them over properly, to say nothing of the cost of making sure people don't just roll back your changes via the passage of time.

Even if 1000 range(1) calls take 50% more time then 1000 view(1) calls the benefits of that change, maybe 0.00001th of a second, will never make up for the amount of time you spend on it. They don't really do much for the time of other coders either (readability) one way or another, outside of some minor oddities like turfs_range - src instead of orange. I respect xeno a ton, they've done some pretty damn cool shit in the past, and they saved me from pipe harddel hell. I'm unwilling to mirror the same devotion to saving every last microsecond that happen on bee.

If we were willing to trade dev time for cpu to that degree we'd make most procs into defines. They're harder to work with, implement and read, but they're significantly faster. I'm unwilling to set this as a standard to meet with current or future work. It is literally not worth your time

I value you and your comfort more then this end result, especially if that time was spend doing jack shit and just chilling out. I wouldn't give it to our most degenerate feature coder, which you certainly aren't.

/towerpost

**Pull request closed without merging**